File: FastAPI/services/schedule_service/services/schedule_service.py
"""
Schedule service implementing business logic.
Includes filtering methods (simplified from Strategy pattern).
"""
from typing import List, Optional
from uuid import UUID
from sqlalchemy.orm import Session
from ..repositories import CourseRepository, ClassRepository
from ..repositories.course_mentor_repository import CourseMentorRepository
from ..models import Course, Class


class ScheduleService:
    """
    Schedule service for managing courses and classes.
    Includes role-based filtering methods.
    """
    
    def __init__(self, db: Session):
        self.db = db
        self.course_repo = CourseRepository(db)
        self.class_repo = ClassRepository(db)
        self.course_mentor_repo = CourseMentorRepository(db)
    
    # ==================== Course Management ====================
    
    def create_course(
        self, 
        code: str, 
        name: str, 
        description: Optional[str] = None,
        mentor_ids: Optional[List[UUID]] = None
    ) -> Course:
        """
        Create a new course with optional mentor assignments.
        
        Args:
            code: Course code
            name: Course name
            description: Course description
            mentor_ids: List of mentor UUIDs to assign
            
        Returns:
            Created course
        """
        # Check if code already exists
        if self.course_repo.code_exists(code):
            raise ValueError(f"Course code '{code}' already exists")
        
        course = Course(code=code, name=name, description=description)
        created_course = self.course_repo.create(course)
        
        # Assign mentors if provided
        if mentor_ids:
            self.course_mentor_repo.set_mentors_for_course(created_course.id, mentor_ids)
        
        return created_course
    
    def get_course(self, course_id: UUID) -> Optional[Course]:
        """Get course by ID."""
        return self.course_repo.find_by_id(course_id)
    
    def get_all_courses(self, skip: int = 0, limit: int = 100) -> List[Course]:
        """Get all courses with pagination."""
        return self.course_repo.find_all(skip=skip, limit=limit)
    
    def update_course(self, course_id: UUID, mentor_ids: Optional[List[UUID]] = None, **kwargs) -> Optional[Course]:
        """
        Update course.
        
        Args:
            course_id: UUID of the course
            mentor_ids: List of mentor UUIDs to assign (replaces existing)
            **kwargs: Fields to update
            
        Returns:
            Updated course
        """
        # If updating code, check for duplicates
        if 'code' in kwargs:
            existing = self.course_repo.find_by_code(kwargs['code'])
            if existing and existing.id != course_id:
                raise ValueError(f"Course code '{kwargs['code']}' already exists")
        
        # Update mentor assignments if provided
        if mentor_ids is not None:
            self.course_mentor_repo.set_mentors_for_course(course_id, mentor_ids)
        
        return self.course_repo.update(course_id, **kwargs)
    
    def delete_course(self, course_id: UUID) -> bool:
        """
        Delete course.
        
        Args:
            course_id: UUID of the course
            
        Returns:
            True if deleted
        """
        return self.course_repo.delete(course_id)
    
    # ==================== Course Mentor Management ====================
    
    def get_mentors_for_course(self, course_id: UUID) -> List[UUID]:
        """Get all mentor IDs assigned to a course."""
        return self.course_mentor_repo.get_mentors_for_course(course_id)
    
    def get_courses_for_mentor(self, mentor_id: UUID) -> List[UUID]:
        """Get all course IDs a mentor is assigned to."""
        return self.course_mentor_repo.get_courses_for_mentor(mentor_id)
    
    def assign_mentor_to_course(self, course_id: UUID, mentor_id: UUID) -> bool:
        """Assign a mentor to a course."""
        self.course_mentor_repo.assign_mentor(course_id, mentor_id)
        return True
    
    def remove_mentor_from_course(self, course_id: UUID, mentor_id: UUID) -> bool:
        """Remove a mentor from a course."""
        return self.course_mentor_repo.remove_mentor(course_id, mentor_id)
    
    def is_mentor_assigned_to_course(self, course_id: UUID, mentor_id: UUID) -> bool:
        """Check if a mentor is assigned to a course."""
        return self.course_mentor_repo.is_mentor_assigned_to_course(course_id, mentor_id)
    
    # ==================== Class Management ====================
    
    def create_class(
        self,
        course_id: UUID,
        name: str,
        room_number: str,
        day_of_week: str,
        schedule_time,
        mentor_id: Optional[UUID] = None
    ) -> Class:
        """
        Create a new class.
        
        Args:
            course_id: UUID of the course
            name: Class name
            room_number: Room number
            day_of_week: Day of the week
            schedule_time: Class time
            mentor_id: UUID of the mentor (optional)
            
        Returns:
            Created class
        """
        # Verify course exists
        if not self.course_repo.exists(course_id):
            raise ValueError(f"Course with ID {course_id} does not exist")
        
        class_obj = Class(
            course_id=course_id,
            mentor_id=mentor_id,
            name=name,
            room_number=room_number,
            day_of_week=day_of_week,
            schedule_time=schedule_time
        )
        
        return self.class_repo.create(class_obj)
    
    def get_class(self, class_id: UUID) -> Optional[Class]:
        """Get class by ID."""
        return self.class_repo.find_by_id(class_id)
    
    def get_all_classes(self, skip: int = 0, limit: int = 100) -> List[Class]:
        """Get all classes with pagination."""
        return self.class_repo.find_all(skip=skip, limit=limit)
    
    def update_class(self, class_id: UUID, **kwargs) -> Optional[Class]:
        """
        Update class.
        
        Args:
            class_id: UUID of the class
            **kwargs: Fields to update
            
        Returns:
            Updated class
        """
        return self.class_repo.update(class_id, **kwargs)
    
    def delete_class(self, class_id: UUID) -> bool:
        """
        Delete class.
        
        Args:
            class_id: UUID of the class
            
        Returns:
            True if deleted
        """
        return self.class_repo.delete(class_id)
    
    # ==================== Schedule Filtering ====================
    
    def get_schedule_for_student(self, student_id: UUID, skip: int = 0, limit: int = 100) -> List[Class]:
        """
        Get schedule for a specific student (only enrolled classes).
        """
        return self.class_repo.find_by_student(student_id, skip=skip, limit=limit)
    
    def get_schedule_for_mentor(self, mentor_id: UUID, skip: int = 0, limit: int = 100) -> List[Class]:
        """
        Get schedule for a specific mentor (only assigned classes).
        """
        return self.class_repo.find_by_mentor(mentor_id, skip=skip, limit=limit)
    
    def get_full_schedule(self, skip: int = 0, limit: int = 100) -> List[Class]:
        """
        Get full schedule (all classes) - for admins/supervisors.
        """
        return self.class_repo.find_all(skip=skip, limit=limit)
    
    def get_schedule_by_role(
        self,
        user_id: UUID,
        role: str,
        skip: int = 0,
        limit: int = 100
    ) -> List[Class]:
        """
        Get schedule based on user role.
        """
        if role == 'student':
            return self.get_schedule_for_student(user_id, skip, limit)
        elif role == 'mentor':
            return self.get_schedule_for_mentor(user_id, skip, limit)
        elif role == 'admin':
            return self.get_full_schedule(skip, limit)
        else:
            raise ValueError(f"Invalid role: {role}")
    
    # ==================== Additional Queries ====================
    
    def get_classes_by_day(self, day: str) -> List[Class]:
        """Get all classes for a specific day."""
        return self.class_repo.find_by_day(day)
    
    def get_classes_by_room(self, room_number: str) -> List[Class]:
        """Get all classes in a specific room."""
        return self.class_repo.find_by_room(room_number)
    
    def get_classes_by_course(self, course_id: UUID) -> List[Class]:
        """Get all classes for a specific course."""
        return self.class_repo.find_by_course(course_id)
    
    # ==================== Conflict Checking ====================
    
    def check_class_conflicts(
        self,
        room_number: str,
        day_of_week: str,
        schedule_time,
        mentor_id: Optional[UUID] = None,
        duration_minutes: int = 90,
        exclude_class_id: Optional[UUID] = None
    ) -> dict:
        """
        Check for scheduling conflicts before creating/updating a class.
        
        Returns:
            dict with 'room_conflicts' and 'mentor_conflicts' lists
        """
        conflicts = {
            'room_conflicts': [],
            'mentor_conflicts': [],
            'has_conflicts': False
        }
        
        # Check room conflicts
        room_conflicts = self.class_repo.find_conflicts_by_room(
            room_number=room_number,
            day_of_week=day_of_week,
            schedule_time=schedule_time,
            duration_minutes=duration_minutes,
            exclude_class_id=exclude_class_id
        )
        
        if room_conflicts:
            conflicts['room_conflicts'] = [
                {
                    'id': str(c.id),
                    'name': c.name,
                    'room_number': c.room_number,
                    'day_of_week': c.day_of_week,
                    'schedule_time': str(c.schedule_time),
                    'course_name': c.course.name if c.course else None
                }
                for c in room_conflicts
            ]
            conflicts['has_conflicts'] = True
        
        # Check mentor conflicts if mentor is specified
        if mentor_id:
            mentor_conflicts = self.class_repo.find_conflicts_by_mentor(
                mentor_id=mentor_id,
                day_of_week=day_of_week,
                schedule_time=schedule_time,
                duration_minutes=duration_minutes,
                exclude_class_id=exclude_class_id
            )
            
            if mentor_conflicts:
                conflicts['mentor_conflicts'] = [
                    {
                        'id': str(c.id),
                        'name': c.name,
                        'room_number': c.room_number,
                        'day_of_week': c.day_of_week,
                        'schedule_time': str(c.schedule_time),
                        'course_name': c.course.name if c.course else None
                    }
                    for c in mentor_conflicts
                ]
                conflicts['has_conflicts'] = True
        
        return conflicts
    
    def create_class_with_validation(
        self,
        course_id: UUID,
        name: str,
        room_number: str,
        day_of_week: str,
        schedule_time,
        mentor_id: Optional[UUID] = None,
        duration_minutes: int = 90
    ) -> dict:
        """
        Create a class with conflict validation.
        
        Returns:
            dict with 'success', 'class' (if created), and 'conflicts' (if any)
        """
        # Check for conflicts first
        conflicts = self.check_class_conflicts(
            room_number=room_number,
            day_of_week=day_of_week,
            schedule_time=schedule_time,
            mentor_id=mentor_id,
            duration_minutes=duration_minutes
        )
        
        if conflicts['has_conflicts']:
            return {
                'success': False,
                'class': None,
                'conflicts': conflicts
            }
        
        # No conflicts, create the class
        class_obj = self.create_class(
            course_id=course_id,
            name=name,
            room_number=room_number,
            day_of_week=day_of_week,
            schedule_time=schedule_time,
            mentor_id=mentor_id
        )
        
        return {
            'success': True,
            'class': class_obj,
            'conflicts': None
        }

File: frontend/my-app/src/services/scheduleService.ts
/**
 * Schedule Service - API calls for courses, classes, and enrollments
 */
import api from './api';

// Types
export interface MentorInfo {
  id: string;
  full_name: string;
  email: string;
}

export interface Course {
  id: string;
  code: string;
  name: string;
  description?: string;
  mentor_ids: string[];
  mentors: MentorInfo[];
  created_at: string;
  updated_at: string;
}

export interface CourseCreate {
  code: string;
  name: string;
  description?: string;
  mentor_ids?: string[];
}

export interface Class {
  id: string;
  course_id: string;
  mentor_id?: string;
  name: string;
  room_number: string;
  day_of_week: string;
  schedule_time: string;
  state: 'active' | 'inactive' | 'completed';
  created_at: string;
  updated_at: string;
  course?: Course;
}

export interface ClassCreate {
  course_id: string;
  mentor_id?: string;
  name: string;
  room_number: string;
  day_of_week: string;
  schedule_time: string;
}

export interface Enrollment {
  id: string;
  student_id: string;
  student_name?: string;
  student_readable_id?: string;
  class_id: string;
  enrolled_at: string;
}

export interface EnrollmentCreate {
  student_id: string;
  class_id: string;
}

// Conflict checking types
export interface ClassConflict {
  id: string;
  name: string;
  room_number: string;
  day_of_week: string;
  schedule_time: string;
  course_name?: string;
}

export interface ConflictCheckResult {
  room_conflicts: ClassConflict[];
  mentor_conflicts: ClassConflict[];
  has_conflicts: boolean;
}

export interface CreateWithValidationResult {
  success: boolean;
  class: Class | null;
  conflicts: ConflictCheckResult | null;
}

// Course API
export const courseApi = {
  getAll: (skip = 0, limit = 100) =>
    api.get<Course[]>('/api/schedule/courses', { params: { skip, limit } }),

  getById: (id: string) =>
    api.get<Course>(`/api/schedule/courses/${id}`),

  create: (data: CourseCreate) =>
    api.post<Course>('/api/schedule/courses', data),

  update: (id: string, data: Partial<CourseCreate>) =>
    api.put<Course>(`/api/schedule/courses/${id}`, data),

  delete: (id: string) =>
    api.delete(`/api/schedule/courses/${id}`),

  // Course mentor management
  getMentors: (courseId: string) =>
    api.get<MentorInfo[]>(`/api/schedule/courses/${courseId}/mentors`),

  assignMentor: (courseId: string, mentorId: string) =>
    api.post(`/api/schedule/courses/${courseId}/mentors`, { mentor_id: mentorId }),

  removeMentor: (courseId: string, mentorId: string) =>
    api.delete(`/api/schedule/courses/${courseId}/mentors/${mentorId}`),
};

// Class API
export const classApi = {
  getAll: (skip = 0, limit = 100) =>
    api.get<Class[]>('/api/schedule/classes', { params: { skip, limit } }),

  getById: (id: string) =>
    api.get<Class>(`/api/schedule/classes/${id}`),

  create: (data: ClassCreate) =>
    api.post<Class>('/api/schedule/classes', data),

  update: (id: string, data: Partial<ClassCreate>) =>
    api.put<Class>(`/api/schedule/classes/${id}`, data),

  delete: (id: string) =>
    api.delete(`/api/schedule/classes/${id}`),

  // Conflict checking
  checkConflicts: (params: {
    room_number: string;
    day_of_week: string;
    schedule_time: string;
    mentor_id?: string;
    duration_minutes?: number;
    exclude_class_id?: string;
  }) =>
    api.post<ConflictCheckResult>('/api/schedule/classes/check-conflicts', null, { params }),

  createWithValidation: (data: ClassCreate, duration_minutes = 90) =>
    api.post<CreateWithValidationResult>('/api/schedule/classes/create-with-validation', data, {
      params: { duration_minutes }
    }),

  // Schedule filtering
  getByStudent: (studentId: string, skip = 0, limit = 100) =>
    api.get<Class[]>(`/api/schedule/schedule/student/${studentId}`, { params: { skip, limit } }),

  getByMentor: (mentorId: string, skip = 0, limit = 100) =>
    api.get<Class[]>(`/api/schedule/schedule/mentor/${mentorId}`, { params: { skip, limit } }),

  getFullSchedule: (skip = 0, limit = 100) =>
    api.get<Class[]>('/api/schedule/schedule/full', { params: { skip, limit } }),

  getByDay: (day: string) =>
    api.get<Class[]>(`/api/schedule/schedule/day/${day}`),

  getByRoom: (roomNumber: string) =>
    api.get<Class[]>(`/api/schedule/schedule/room/${roomNumber}`),
};

// Enrollment API
export const enrollmentApi = {
  enroll: (data: EnrollmentCreate) =>
    api.post<Enrollment>('/api/schedule/enrollments', data),

  unenroll: (studentId: string, classId: string) =>
    api.delete(`/api/schedule/enrollments/${studentId}/${classId}`),

  getByStudent: (studentId: string) =>
    api.get<Enrollment[]>(`/api/schedule/enrollments/student/${studentId}`),

  getByClass: (classId: string) =>
    api.get<Enrollment[]>(`/api/schedule/enrollments/class/${classId}`),

  getClassCount: (classId: string) =>
    api.get<number>(`/api/schedule/enrollments/class/${classId}/count`),
};
